File: 572 - Subtree of Another Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} s
 * @param {TreeNode} t
 * @return {boolean}
 */
var isSubtree = function (s, t) {
    if (s == null && t == null) {
        return true;
    }
    if (s == null || t == null) {
        return false;
    }
    return isSametree(s, t) || isSubtree(s.left, t) || isSubtree(s.right, t);
};
function isSametree(s, t) {
    if (s == null && t == null) {
        return true;
    }
    if (s == null || t == null) {
        return false;
    }
    return s.val == t.val && isSametree(s.left, t.left) && isSametree(s.right, t.right);
}

File: 572 - Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        if t is None and s is None:
            return True
        if t is None or s is None:
            return False
        return self.isSametree(s, t) or self.isSubtree(s.left, t) or self.isSubtree(s.right, t)
    def isSametree(self, s, t):
        if t is None and s is None:
            return True
        if t is None or s is None:
            return False
        l = self.isSametree(s.left, t.left)
        r = self.isSametree(s.right, t.right)
        return s.val == t.val and l and r
